### PR TITLE
Build experimental branches with prefix for public review

### LINKED ISSUE #19: Options needed for publishing draft/experiment work as a part of DEFINE
For the work that is ongoing in DEFINE on mapping and shared responsibility/CRM, it would be useful for us to publish the draft reference material so that the public can review without generating the content locally.  It would also encourage use of the branches for evaluation and feedback.

I intended to merge draft metaschema code into the prototype-* branches that I created in OSCAL for mapping and CRM.  I'm not particularly attached to the prefix, but these are not exactly feature branches since they may contain multiple features (merges) for consideration.  If we could choose an appropriate prefix to publish automatically as model drafts or experimental branches, it would be very helpful.

Steps:

- [x] Discuss with team and agree to prefix: agreed to `prototype`.
- [x] Adjust [list_revisions.sh](https://raw.githubusercontent.com/usnistgov/OSCAL-Reference/develop/support/list_revisions.sh) to include branches and/or [generate_modeldoc.sh](https://raw.githubusercontent.com/usnistgov/OSCAL-Reference/develop/support/list_revisions.sh) to build branches with the prefix.
- [x] Adjust the GitHub Actions workflow to support [workflow_dispatch](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows) to interactively call the updated code in GitHub Actions and build updated versions of the site
  - [x] Strech goal: perhaps allow to filter that list in the GitHub interface and intentionally add/remove filters not detected by the script?

### PR DESCRIPTION
**Pull request by aj-stein-nist**
Update scripts and workflows and close #19.

This will update the workflow to allow the following:

1. Override the revisions and not compute them.
1. Override the prefix and ignore overridden revisions.
1. The default, add prototype* as a prefix for publication from the
usnistgov/OSCAL repo.

In the case the last scenario is not desired, a developer with admin
permissions in the repository can provide a '' prefix and all prototype
prefix branches go away and will be silently ignored.

### CODE REVIEW

**Comment by aj-stein-nist** [comment]
@nikitawootten-nist, I am not sure we would want to or not want to adjust the cache keys based on tags as opposed to accompanying revisions, that would make this maybe more complicated, or do I just have to find a way of calling the [updated block in the relevant GitHub Actions block](https://github.com/usnistgov/OSCAL-Reference/blob/d5eac0ba934c301cc537f35d7c5065ffcea83d5e/.github/workflows/pages.yaml#L32)?

File: support/list_revisions.sh
Comment (nikitawootten-nist) [inline support/list_revisions.sh]: B: Scripts such as the release asset generator and modeldoc generator rely on the latest release being last (so that it can be extracted with a simple `./list_revisions.sh | tail -n 1`). The script in its current state will break that behavior.

I would request that for consistency we put the prototype revision extraction in a separate script (or directly in the Makefile). This is why for example the Makefile appends `develop`, as the `list_revisions.sh` script's only concern is extracting tagged versions.

File: .github/workflows/pages.yaml
Comment (nikitawootten-nist) [inline .github/workflows/pages.yaml]: B, maybe nitpicky: To keep things a bit more readable, could we try folding flags into one:
```suggestion
        run: |
          make site -j2 \
            ${{ github.event.inputs.revisions != '' && format("REVISIONS='{0}'", github.event.inputs.revisions) || '' }} \
            ${{ github.event.inputs.revisions == '' && github.event.inputs.prefix && format("PROTOTYPE_BRANCHES_PREFIX='{0}'", github.event.inputs.prefix) || '' }}
```

This has the major benefit of making the default value still controlled by the Makefile. If adopted the `env` block could be removed.

**Review by nikitawootten-nist — CHANGES_REQUESTED** [review]
Some requested changes regarding the repurposing of `list_revisions.sh` and the workflow logic

File: support/list_revisions.sh
Comment (nikitawootten-nist) [reply support/list_revisions.sh]: Maybe part of this PR could be communicating that more clearly, renaming the script to `list_tagged_releases.sh` or something of that sort.

Comment (aj-stein-nist) [reply support/list_revisions.sh]: So that field appends and not prepends `develop`? OK, let's talk on Gitter about how we order the script invocations. I was not aware of the order requirement, I'm sorry. 🤦

File: .github/workflows/pages.yaml
Comment (aj-stein-nist) [reply .github/workflows/pages.yaml]: I am happy with that change. **EDIT:** sorry I now see there might be an issue with that approach, I will message you.

File: support/list_revisions.sh
Comment (nikitawootten-nist) [reply support/list_revisions.sh]: My mistake, the Makefile prepends develop to the `REVISIONS` variable, but both of the generation scripts rely on `list_revisions.sh` to get taggged releases. The order of the `REVISIONS` variable in the Makefile does not matter, but the order of `list_revisions.sh`'s output does.

File: .github/workflows/pages.yaml
Comment (aj-stein-nist) [reply .github/workflows/pages.yaml]: Adjusted per feedback in https://github.com/usnistgov/OSCAL-Reference/pull/23/commits/2a7ced9018dd1a6bc0a40415aa4792ae55a2a682.

File: support/list_revisions.sh
Comment (aj-stein-nist) [reply support/list_revisions.sh]: https://github.com/usnistgov/OSCAL-Reference/pull/23/commits/2a7ced9018dd1a6bc0a40415aa4792ae55a2a682

**Comment by aj-stein-nist** [comment]
> @nikitawootten-nist, I am not sure we would want to or not want to adjust the cache keys based on tags as opposed to accompanying revisions, that would make this maybe more complicated, or do I just have to find a way of calling the [updated block in the relevant GitHub Actions block](https://github.com/usnistgov/OSCAL-Reference/blob/d5eac0ba934c301cc537f35d7c5065ffcea83d5e/.github/workflows/pages.yaml#L32)?

On Monday, @nikitawootten-nist we should make sure we understand how this will work and its caching is logical.

@Arminta-Jenkins-NIST it changed a bit so we should re-review on Monday morning.

**Comment by aj-stein-nist** [comment]
Examples from my fork with `main` aligned to the PR branch for testing "once merged" after Nikita's feedback changes by me were completed.

1. [Running as is with no alternate revisions or branch prefix](https://github.com/aj-stein-nist/OSCAL-Reference/actions/runs/6279027875)
2. [Running with targeted revisions, skipping one of the two prototype branches](https://github.com/aj-stein-nist/OSCAL-Reference/actions/runs/6279726735) in the revisions argument set to `develop v1.0.0 v1.0.1 v1.0.2 v1.0.3 v1.0.4 v1.0.5 v1.0.6 v1.1.0 v1.1.1 prototype-mapping-model`, excluding `prototype-shared-responsibility-model` from the list so it is "unpublished" on next run)
3. [Running with no revision change but picking a new custom prefix](https://github.com/aj-stein-nist/OSCAL-Reference/actions/runs/6279728061), `does_not_exist`, which will return nothing and unpublish all branches from before with prefix `prototype`

It seems 3 needs to be investigated as the prototype prefix branch still includes the prototype-mapping branch from 2 even though it should not, so maybe this is not a done deal so to speak.

File: .github/workflows/pages.yaml
Comment (nikitawootten-nist) [inline .github/workflows/pages.yaml]:
```suggestion
          echo revisions_hash=$(make --silent list-tags | sha256sum) >> $GITHUB_OUTPUT
```

File: .github/workflows/pages.yaml
Comment (nikitawootten-nist) [inline .github/workflows/pages.yaml]:
```suggestion
          echo branches="$(make --silent list-prototype-branches PROTOTYPE_BRANCHES_PREFIX='$PREFIX')" >> $GITHUB_OUTPUT
```

File: support/list_branches.sh
Comment (nikitawootten-nist) [inline support/list_branches.sh]: lines must be moved **below** the definition of `usage()`

File: .github/workflows/pages.yaml
Comment (aj-stein-nist) [reply .github/workflows/pages.yaml]: Fixed in pairing.

File: support/list_branches.sh
Comment (aj-stein-nist) [reply support/list_branches.sh]: Fixed in pairing.

File: .github/workflows/pages.yaml
Comment (aj-stein-nist) [reply .github/workflows/pages.yaml]: Fixed in pairing.

**Comment by aj-stein-nist** [comment]
OK @nikitawootten-nist, ready for reviewed based on fork testing this appears to be good, and we confirmed my impressions about missed branch deletions with incorrect. We can finalize beginning of next sprint if necessary.

**Comment by aj-stein-nist** [comment]
OK per discussion in sprint planning and pairing this afternoon, we shall exempt the prototype branches from link checking with the expectation once you try to point into OSCAL `develop`, that is the step beyond prototype and you need to check those links before it can get published. That commit has been added on the (now properly) rebased PR branch below.

https://github.com/usnistgov/OSCAL-Reference/pull/23/commits/98405dc857b97de26f08a73d1a6799f726728a9a

**Comment by aj-stein-nist** [comment]
Still seem to be getting some failures and trying to clean them up, for others I am testing locally with `make linkcheck LYCHEE_EXTRA_FLAGS="--exclude-path=*/models/prototype*" REVISIONS="develop $(make -s list-prototype-branches)"` to mimic the implied expansion of shell commands from the Makefile and GHA defaults.

**Comment by aj-stein-nist** [comment]
I will need to finish examining later, but seems that somehow when testing locally the bad `mapping` links I have from building develop and prototype branches is in develop, not the prototype branches. 🤷

https://gist.github.com/aj-stein-nist/8a1ae4ab17f57ef9ffcb6022248e9439

I will make clean for the third or fourth time after rebasing with Nikita on the call, but ... odd?

**Comment by aj-stein-nist** [comment]
Since I've been reading logs a lot, something I need to consider fixing in this PR or in a follow-up one.

```sh
[WARN ] WARNING: `--exclude-file` is deprecated and will soon be removed; use `.lycheeignore` file to ignore URL patterns instead. To exclude paths of files and directories, use `--exclude-path`.
```

**Comment by aj-stein-nist** [comment]
OK, @nikitawootten-nist and others, it turns out that the only thing breaking now is timeouts on a itu.int link that has been working in my browser but even failed intermittently with links and such. I resolved the previous issue by getting rid of the regex wildcard technique for use with `lychee` to ignore paths to prototype branches (and not develop) for links that are not prime-time ready yet. So what's below didn't work.

```sh
lychee --exclude-file ./support/lychee_ignore.txt  \
  --config ./support/lychee.toml \
  --output lychee_report.md \
  --verbose --format markdown \
  --exclude-path="*public/models*" \
  site/public/**/*.html
```
I tried many, many different paths, then even specific paths with the wildcards before the path, after it, and both, with `public/develop` or `develop` and it still reported bad links checks there, so something isn't right there. Before we were using GHA workflows to stub in the branch name between wildcards, but I blew that way, when I realized some of the errors I was adding on purpose could not be silenced in the key places, so this approach was a DOA.

So I pivoted to using the list of branches to build a list of `--exclude-path=site/public/models/branchname1 --exclude-path=site/public/models/branchname2` to the extra arguments override in the Makefile.

```sh
lychee --exclude-file ./support/lychee_ignore.txt  \
  --config ./support/lychee.toml \
  --output lychee_report.md \
  --verbose --format markdown \
  --exclude-path=site/public/models/prototype-mapping-model \
  --exclude-path=site/public/models/prototype-shared-responsibility-model \
  site/public/**/*.html
```
Then we use the GitHub Actions workflow to pass in the orchestrated by the pages.yaml workflow by using the Linux/Unix utility `printf` to take that list of branches and create the stringified form of that one branch name at a time. The `\055` is the encoding of the first `-` in `--` for `--exclude-path` because `printf` sees that as an argument and you must be careful about that, hence the Stackoverflow note. The relevant argument for others to understand what I mean is `linkcheck_exclude_path_arg=$(...)`.

```yaml
# Truncated other irrelevant code
      - name: Get the list of tagged revisions (for cache)
        id: get-revisions
        run: |
          echo revisions_hash=$(make --silent list-tags | sha256sum) >> $GITHUB_OUTPUT
        shell: bash
      - name: Get the list of prototype branches (for cache)
        id: get-prototype-branches
        run: |
          {
            # Wildcards after a stable tag prefix like v* for v1.0.0 will not work, so generate outputs for content and data paths for next step
            branches="$(make --silent list-prototype-branches PROTOTYPE_BRANCHES_PREFIX=$PREFIX)"
            echo branches="$branches"
            [[ -n "${branches-}" ]] && echo content_paths="$(printf "site/content/models/%s " $branches)"
            [[ -n "${branches-}" ]] && echo data_paths="$(printf "site/content/data/%s " $branches)"
            # Be forewarned, printf -- has a special meaning so we must encode the first - as \055 in -- to avoid errors
            # Explanation: https://unix.stackexchange.com/a/658388
            [[ -n "${branches-}" ]] && echo linkcheck_exclude_path_arg=$(printf "\055-exclude-path='site/public/models/%s' "  $branches)
            # Prototype branch content may change so hash branch name and current commit hash for its HEAD as restore key part
            echo branches_hash="$(cd "${OSCAL_DIR}"; git ls-remote origin $PREFIX | sha256sum | cut -d ' ' -f 1)"
          } >> $GITHUB_OUTPUT
# Truncated other irrelevant code
      - name: Link Check
        id: linkcheck
        run: |
          make linkcheck LYCHEE_EXTRA_FLAGS='--github-token ${{ secrets.GITHUB_TOKEN }} \
            ${{ steps.get-prototype-branches.outputs.linkcheck_exclude_path_arg && format('{0}', steps.get-prototype-branches.outputs.linkcheck_exclude_path_arg) || '' }}'
```
I will keep re-running, but I think it works for now. I am working on a bug report to the lychee folk, but this is ready for review and merge I think.

File: .github/workflows/pages.yaml
Comment (nikitawootten-nist) [inline .github/workflows/pages.yaml:87]: NB: That sounds like it was painful to debug, nice catch

File: Makefile
Comment (nikitawootten-nist) [inline Makefile]: NB: Suggest adding more context for future maintainers here:
```suggestion
# `REVISIONS` contains a list of Git refs (specifically branches and tags) of the OSCAL repository that documentation will be built for.
# By default, `REVISIONS` includes:
# 1. The `develop` branch, showcasing changes staged for the next release.
# 2. Tagged releases matching a semver pattern (e.g. `v.1.0.0`).
# 3. Prototype branches matching the specified `PROTOTYPE_BRANCHES_REMOTE` and `PROTOTYPE_BRANCHES_PREFIX`.
# For local development, `REVISIONS` can be overridden to build a subset of the site
#   (e.g. `make site REVISIONS='v1.1.0 my-special-branch'`)
```

File: support/list_branches.sh
Comment (nikitawootten-nist) [inline support/list_branches.sh]: NB: Suggest clarifying the purpose of this script
```suggestion
# List all OSCAL branches from the specified git remote matching the specified prefix (e.g. prototype)
```

**Review by nikitawootten-nist — APPROVED** [review]
Minor NB doc suggestions below, looks good!

File: .github/workflows/pages.yaml
Comment (aj-stein-nist) [reply .github/workflows/pages.yaml:87]: Well, I got through it! 😅

**Pull request merged**
